File: frontend/src/lib/spatialCoordinates.ts
/**
 * Raumvertrag — muss mit backend/app/core/spatial_coordinates.py übereinstimmen.
 */

import type { HeightMatrix } from "@/lib/isoBuilding";

export const SPATIAL_GAP = 1.02;

export const SPATIAL_COORDINATE_SYSTEM = {
  frontRow: 0,
  xDirection: "left-to-right" as const,
  yDirection: "front-to-back" as const,
  zDirection: "bottom-to-top" as const,
};

export type SpatialCameraPreset =
  | "oblique"
  | "front"
  | "back"
  | "left"
  | "right"
  | "top"
  | "front_left"
  | "front_right"
  | "back_left"
  | "back_right";

export type BuildingSceneBounds = {
  cx: number;
  cy: number;
  cz: number;
  width: number;
  depth: number;
  height: number;
};

export function buildingSceneBounds(matrix: HeightMatrix): BuildingSceneBounds {
  const cols = matrix[0]?.length ?? 1;
  const rows = matrix.length;
  let maxH = 1;
  for (const row of matrix) {
    for (const h of row) {
      if (h > maxH) maxH = h;
    }
  }
  const width = Math.max(SPATIAL_GAP, (cols - 1) * SPATIAL_GAP + 0.94);
  const depth = Math.max(SPATIAL_GAP, (rows - 1) * SPATIAL_GAP + 0.94);
  const height = maxH * SPATIAL_GAP;
  return {
    cx: 0,
    cy: height / 2,
    cz: 0,
    width,
    depth,
    height,
  };
}

File: backend/app/core/spatial_coordinates.py
"""Einheitlicher Raumvertrag für Höhenmatrix, Three.js und Ansichten.

Alle Module (iso_building, spatial_compact, Frontend spatialCoordinates) beziehen sich auf:

- matrix[row][col] = matrix[y][x] — Zeile 0 = **vorne** (niedriges y / niedriges gy im 3D-Modell).
- x wächst von links nach rechts (Spalte 0 = links).
- z (gz) = Würfelhöhe von unten nach oben (Wert in der Zelle = Anzahl übereinanderliegender Würfel).

Three.js (VoxelBuilding): position (gx * GAP, gz * GAP, gy * GAP) im lokalen Gebäude-Group,
nach Zentrierung: Vorne ≈ negative Welt-Z-Richtung vom Betrachter aus «von vorne».

Orthographische Lehrbuch-Ansichten (building_projections):
- top: Grundriss mit Höhenzahlen pro Feld.
- front: Blick von vorne (−y / −Z); pro Spalte x die Silhouettenhöhe max_y matrix[y][x].
- right: Blick von rechts (+x); pro Tiefe y die Silhouettenhöhe max_x matrix[y][x].
"""

from __future__ import annotations

from typing import TypedDict

FRONT_ROW: int = 0

SPATIAL_COORDINATE_SYSTEM: dict[str, str] = {
    "front_row": "0",
    "x_direction": "left-to-right",
    "y_direction": "front-to-back",
    "z_direction": "bottom-to-top",
}


class SpatialCoordinateSystem(TypedDict):
    front_row: int
    x_direction: str
    y_direction: str
    z_direction: str


def spatial_coordinate_system() -> SpatialCoordinateSystem:
    return {
        "front_row": FRONT_ROW,
        "x_direction": "left-to-right",
        "y_direction": "front-to-back",
        "z_direction": "bottom-to-top",
    }
